autometa/common/external/samtools.py

```diff
@@ -26,58 +26,65 @@
 import logging
 import os
 import subprocess
+import shutil
+import tempfile
 
 import pandas as pd
 import multiprocessing as mp
 
 logger = logging.getLogger(__name__)
 
 
-def run(cmd):
-    """Run `cmd` via subprocess.
-
+def run(cmd, bam):
+    """
+    Run `cmd` via subprocess.
+    
     Parameters
     ----------
     cmd : str
         Executable input str
-
+    bam : str
+        </path/to/output/alignment.bam>
+    
     Returns
     -------
     bool
-        True if retcode == 0. i.e. Success.
-        False if retcode != 0. i.e. Failed.
-
+        True if retcode == 0. i.e. function run successfully.
+    
     Raises
-    -------
-    ExceptionName
-        Why the exception is raised.
-
+    ------
+    ChildProcessError
+        Function did not run successfully, returns retcode.
     """
+    
     logger.debug(f'cmd: {cmd}')
-    with open(os.devnull, 'w') as stdout, open(os.devnull, 'w') as stderr:
+    log_samtools_dir = os.path.dirname(bam)
+    tempdir = tempfile.mkdtemp(suffix=None, prefix='samtools', dir=log_samtools_dir)
+    samtools_stderr=os.path.join(tempdir, 'samtools_stderr')
+    samtools_stdout=os.path.join(tempdir, 'samtools_stdout')
+    with open(samtools_stdout, 'w') as stdout, open(samtools_stderr, 'w') as stderr:
         retcode = subprocess.call(cmd, stdout=stdout, stderr=stderr, shell=True)
-    if not retcode:
+        if retcode != 0:
+            raise ChildProcessError(f'Samtools not successfully run, retcode: {retcode}')
+        shutil.rmtree(tempdir, ignore_errors=True)
```

Comment:
Following lines 63-64 above, Perhaps it would be better to remove all intermediate files (whether successful or not) and keep the logs only if the process fails. If this step continues to fail, the alignment files could take up a lot of unnecessary space, where as the logs should not take up too much space.